Improve `NonStaticMagicMethods` sniff (code review).

jrfnl opened the pull request:
* Also sniff `trait`s for magic methods.
* Do the method name comparison in a case-insensitive manner as functions and methods in PHP are handled case-insensitively.
* Changed the array format to allow for checking visibility and static requirements based on the requirements of individual functions.
* Added a number of extra PHP magic methods to the array.

   _For `__sleep()` and `__toString()` only the visibility is explicitly documented, for `__set_state()` only the fact that it has to be static. So currently only checking the documented requirements._
* Added the `SoapClient` magic methods to the array.

   _It does not seem to be explicitly documented whether these have a requirement of being non-static, so that's currently not checked._
* For the `SoapClient` methods no extra check is made to verify whether or not the current class extends `SoapClient` as all methods with a double underscore prefix are reserved for PHP anyway.
* Defer to PHPCS native methods to determine the method name and properties.
* As the messages will be used now to enforce both 'must be static/public' as well as the opposite, changed the message format to be more flexible.
* Removed the _'(since PHP 5.3)'_ part of the error message.
* Removed the 'all-in-one' message as with the extra message flexibility, this would become a quite complex message to create. (Can be brought back if you think it's important to have it)
* Overhauled the unit tests to reflect the above changes.
* Added additional unit tests for traits and for the additional magic methods added (both PHP and SoapClient)

### Open questions

**Regarding the version check**:
As far as I know, the visibility and static requirements of these methods did not change in PHP 5.3. It's just that PHP started throwing warnings about it from that version upwards.
So I'm tempted to remove the check for PHP 5.3+ altogether. Opinions ?

**Regarding the specific property checks**:
* Anyone knows whether the `__sleep()` and `__toString()` method have a requirement to be non-static ?
* Anyone knows whether the `__set_state()` method has a visibility requirement ?
* Anyone knows whether the `SoapClient` magic methods have a requirement regarding them being static/non-static ? I believe these should all be non-static as the native class has them as non-static, but would like a second opinion.

----

Similar to other sniffs, the name of this sniff now no longer covers the functionality and the sniff should be renamed. I'd suggest `MagicMethodPropertyRequirements` as an alternative name.

wimg commented:
http://php.net/manual/en/migration53.incompatible.php mentions : "The magic methods __get(), __set(), __isset(), __unset(), and __call() must always be public and can no longer be static. Method signatures are now enforced."
So I'm guessing that if we remove the version check, unit tests will fail on 5.2 (which we're not running on Travis, I know, but that's the reason the version check was there if I remember correctly).

Not sure about the way the SoapClient methods are implemented, since the sniff doesn't check the inheritance on that class right now ?

jrfnl commented:
http://php.net/manual/en/language.oop5.overloading.php#object.call says:
> Added warning to enforce public visibility and non-static declaration. 

More than anything, the warning was added in 5.3. Whether the requirement actually changed is still not completely clear to me.

> So I'm guessing that if we remove the version check, unit tests will fail on 5.2

Now, they won't as these unit tests are not run against a specific version.

> Not sure about the way the SoapClient methods are implemented, since the sniff doesn't check the inheritance on that class right now ?

True as checking the inheritance would only work for child classes, not grand-children, so is inherently unreliable.

jrfnl commented:
I've made a minor change to the unit tests & recommitted as there was a fail on Travis.

If this sniff was run on the combination PHPCS 1.x + PHP 5.3 or less, the unit tests were failing because traits are not yet recognized properly in PHPCS in that combi.

I've now split the test case file in two - one for the classes and interfaces, one for the traits and added test skip indicators for that specific combination only - PHPCS 1.x + PHP 5.3 or less + test deals with traits.

jrfnl commented:
Maybe I should actually remove the SoapClient methods altogether as they are inherently magic and therefore should just abide by the same rules as any method being overloaded from a child class.

jrfnl commented:
Another thing - as the sniff is now more flexible -, we could also include a selective check for magic methods which throw exceptions as there are a number of magic methods from which this is not allowed.

wimg commented:
1 more remark : could you rename the tests files so they indicate their contents instead of just 1 and 2 ? Nobody will remember what the difference is in a few weeks ;-)

jrfnl commented:
Done - Renamed 1 to just the sniff file as 'normal' tests should be added there. Suffixed `.._traits` to the second one to make it clear that file only deals with traits.
Oh and the reason for the split is (and was) documented in the test file itself where the files are being read in.

jrfnl commented:
Oh, just to be sure: the test files are still called `nonstatic_magic_methods....` to be in line with the sniff name. So this doesn't yet address the fact that the sniff now covers more than the name implies.
